Products/CMFPlone/browser/login/login.py

```diff
@@ -115,19 +115,35 @@ def updateWidgets(self):
         self.widgets["came_from"].value = self.get_came_from()
 
     def get_came_from(self):
+        """Get and validate the came_from parameter for redirect after login."""
         came_from = self.request.get("came_from", None)
         if not came_from:
             came_from = self.request.get("HTTP_REFERER", None)
             if not came_from:
-                return
+                return None
+                
         url_tool = getToolByName(self.context, "portal_url")
+        
+        # First check if URL is in portal
         if not url_tool.isURLInPortal(came_from):
-            return
-        came_from_path = parse.urlparse(came_from)[2].split("/")
-        for login_template_id in LOGIN_TEMPLATE_IDS:
-            if login_template_id in came_from_path:
-                return
-        return came_from
+            return None
+            
+        try:
```

Comment:
Please revert your chances in `get_came_from` completely ... there is no need for this try/except here.